stl/inc/format

```diff
@@ -0,0 +1,2734 @@
+        _Specs._Width = _Width;
+    }
+
+    constexpr void _On_precision(int _Precision) {
+        _Specs._Precision = _Precision;
+    }
+
+    constexpr void _On_localized() {
+        _Specs._Localized = true;
+    }
+
+    constexpr void _On_type(_CharT _Type) {
+        // performance note: this could be optimized to one comparison by
+        // first casting to unsigned int (the negative values will be 128-255)
+        if (_Type < 0 || _Type > (numeric_limits<signed char>::max)()) {
+            throw format_error("Invalid type specification.");
+        }
+        _Specs._Type = static_cast<char>(_Type);
+    }
+
+protected:
+    _Basic_format_specs<_CharT>& _Specs;
+};
+
+template <class _Context>
+_NODISCARD constexpr basic_format_arg<_Context> _Get_arg(const _Context& _Ctx, size_t _Arg_id) {
+    // note: while this is parameterized on the _Arg_id type in libfmt we don't
+    // need to do that in std::format because it's only called with either an integer
+    // id or a named id (which we do not support in std::format)
+    auto _Arg = _Ctx.arg(_Arg_id);
+    if (!_Arg) {
+        throw format_error("Argument not found.");
+    }
+    return _Arg;
+}
+
+// Checks that the type and value of an argument associated with a dynamic
+// width specifier are valid.
+class _Width_checker {
+public:
+    template <class _Ty>
+    _NODISCARD constexpr unsigned long long operator()(_Ty _Value) {
+        if constexpr (is_integral_v<_Ty>) {
+            if constexpr (is_signed_v<_Ty>) {
+                if (_Value < 0) {
+                    throw format_error("Negative width.");
+                }
+            }
+            return static_cast<unsigned long long>(_Value);
+        } else {
+            throw format_error("Width is not an integer.");
+        }
+    }
+};
+
+// Checks that the type and value of an argument associated with a dynamic
+// precision specifier are valid.
+class _Precision_checker {
+public:
+    template <class _Ty>
+    _NODISCARD constexpr unsigned long long operator()(_Ty _Value) {
+        if constexpr (is_integral_v<_Ty>) {
+            if constexpr (is_signed_v<_Ty>) {
+                if (_Value < 0) {
+                    throw format_error("Negative precision.");
+                }
+            }
+            return static_cast<unsigned long long>(_Value);
+        } else {
+            throw format_error("Precision is not an integer.");
+        }
+    }
+};
+
+// Fetch the value of an argument associated with a dynamic
+// width or precision specifier. This will be called with either
+// _Width_checker or _Precision_checker as "_Handler".
+template <class _Handler, class _FormatArg>
+_NODISCARD constexpr int _Get_dynamic_specs(const _FormatArg _Arg) {
+    _STL_INTERNAL_STATIC_ASSERT(_Is_any_of_v<_Handler, _Width_checker, _Precision_checker>);
+    const unsigned long long _Val = _STD visit_format_arg(_Handler(), _Arg);
+    if (_Val > (numeric_limits<int>::max)()) {

```

Comment:
This compares `unsigned long long` with `int` Should we cast?

Reply:
Although the compilers don't seem to be warning here (and the specific value is preserved), mixing signed and unsigned is always scary, so a cast is reasonable.

Reply:
I think the cleaner solution would be to make `_Val` an int, as that is what we are returning